### PR TITLE
Add a GitHub document loader

### LINKED ISSUE #411: [FEATURE] Add a GitHub document loader
LangChain4J currently has an AWS S3 document loader at https://github.com/langchain4j/langchain4j/tree/main/document-loaders/langchain4j-document-loader-amazon-s3

In order to load content from GitHub, I would like to add a GitHub document loader (as https://js.langchain.com/docs/integrations/document_loaders/web_loaders/github for JavaScript).

My ultimate goal is to be able to use the RAG pattern with GitHub repositories.

### PR DESCRIPTION
**Pull request by jdubois**
Fix #411

### CODE REVIEW

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+import org.kohsuke.github.GitHub;
+import org.kohsuke.github.GitHubBuilder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GithubDocumentLoader {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: maybe `Github`->`GitHub`? Please check everywhere that `gitHubXxx` is written consistently

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+    private static final Logger logger = LoggerFactory.getLogger(GithubDocumentLoader.class);
+
+    private String apiUrl = "https://api.github.com";
+
+    private final GitHub github;
+
+    public GithubDocumentLoader(String apiUrl, String githubToken, String gitHubTokenOrganization) throws IOException {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: I would suggest wrapping and rethrowing `IOException` for a better dev experience, in other places as well

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    public Document loadDocument(String owner, String repo, String branch, String path, DocumentParser parser) throws IOException {
+        GHContent content = github
+                .getRepository(owner + "/" + repo)
+                .getFileContent(path, branch);
+
+        return fromGitHub(parser, content);
+    }
+
+    public Document loadDocument(String owner, String repo, String branch, DocumentParser parser) throws IOException {
+        return loadDocument(owner, repo, branch,"", parser);
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: What is the purpose of this method? Could you please add a test for this case?

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    private static Document fromGitHub(DocumentParser parser, GHContent content) {
+        logger.info("Loading document from GitHub: {}", content.getHtmlUrl());
+        try {
+            if (content.isFile()) {
+                GithubSource source = new GithubSource(content);
+                return DocumentLoader.load(source, parser);
+            } else {
+                logger.debug("Skipping directory: {}", content.getHtmlUrl());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: This branch can be executed only when called from `loadDocument(String owner, String repo, String branch, String path, DocumentParser parser)` (when user specifies `path` explicitly). In case user provided path that is not a file, I think we should throw an `IllegalArgumentException` instead of returning null and logging.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static class Builder {
+
+        private String apiUrl = "https://api.github.com";
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: no need to define it here, it is already defined in the github library

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        }
+
+        public Builder gitHubTokenOrganization(String gitHubTokenOrganization) {
+            this.gitHubTokenOrganization = gitHubTokenOrganization;
+            return this;
+        }
+
+        public GithubDocumentLoader build() throws IOException {
+            if (github != null) {
+                return new GithubDocumentLoader(github);
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: To simplify, I would suggest removing this case from the builder and just use existing public ctor that accepts `GitHub` instance

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    public Metadata metadata() {
+        Metadata metadata = new Metadata();
+        metadata.add("git_url", content.getGitUrl());
+        try {
+            metadata.add("download_url", content.getDownloadUrl());
+        } catch (IOException e) {
+            // Ignore if download_url is not available
+        }
+        metadata.add("html_url", content.getHtmlUrl());
+        metadata.add("name", content.getName());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java]: what do you think about prepending all keys with "github_" prefix and making key name more specific (e.g. "github_file_name", "github_file_size", etc)?

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+public class GithubDocumentLoader {
+
+    private static final Logger logger = LoggerFactory.getLogger(GithubDocumentLoader.class);
+
+    private String apiUrl = "https://api.github.com";
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: can be removed

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+
+    private final GitHub github;
+
+    public GithubDocumentLoader(String apiUrl, String githubToken, String gitHubTokenOrganization) throws IOException {
+        GitHubBuilder gitHubBuilder = new GitHubBuilder();
+        if (apiUrl != null) {
+            this.apiUrl = apiUrl;
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: minor: no need in this, you can safely pass null into `withEndpoint()`, it will use `https://api.github.com` if `null` under the hood:
`gitHubBuilder.withEndpoint(apiUrl)`

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                gitHubBuilder.withOAuthToken(githubToken, gitHubTokenOrganization);
+            }
+        }
+        github = gitHubBuilder.build();
+    }
+
+    public GithubDocumentLoader(GitHub github) {
+        this.github = github;
+    }
+
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: Seems that public repositories can be accessed without a token, so maybe we should have a no-arg ctor here as well, WDYT?

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        metadata.add("path", content.getPath());
+        metadata.add("sha", content.getSha());
+        metadata.add("size", Long.toString(content.getSize()));
+        metadata.add("type", content.getType());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java]: As far as I understand, type is always `file`? If yes, we can remove it.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            if (content.isFile()) {
+                GithubSource source = new GithubSource(content);
+                return DocumentLoader.load(source, parser);
+            } else {
+                logger.debug("Skipping directory: {}", content.getHtmlUrl());
+                return null;
+            }
+        } catch (IOException e) {
+            logger.error("Failed to load document from GitHub: {}", content.getHtmlUrl(), e);
+            return null;
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: I would suggest throwing an exception here and do `withRetry(() -> fromGitHub(parser, content), 3)` when calling it.

File: document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoaderIT.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import dev.langchain4j.data.document.parser.TextDocumentParser;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class GithubDocumentLoaderIT {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoaderIT.java]: Please add `@EnabledIfEnvironmentVariable(named = "GITHUB_TOKEN", matches = ".+")`

**Review by ghost — CHANGES_REQUESTED** [review]
@jdubois thanks a lot! Please check my comments.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java (hunk 1 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: Done

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java (hunk 3 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: Sorry for the previous comment, I got messed up...
It's useless, I've removed it.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java (hunk 7 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GithubSource.java]: It feels just a bit weird for `github_git_url`, but otherwise yes it's more clear, so I've added the prefix.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java (hunk 10 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GithubDocumentLoader.java]: Oh yes the token is optional (it's recommended otherwise you'll be rate limited, but it works for simple projects).
I've added the empty constructor.

**Comment by jdubois** [comment]
@langchain4j thanks for the review! It should be all OK.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+
+import static dev.langchain4j.internal.RetryUtils.withRetry;
+
+public class GitHubDocumentLoader {
+
+    private static final Logger logger = LoggerFactory.getLogger(GitHubDocumentLoader.class);
+
+    private final GitHub gitHub;
+
+    public GitHubDocumentLoader(String gitHubToken, String gitHubTokenOrganization) throws IOException {
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: checked `IOException` is still thrown by many methods and ctors, would be nice to wrap it into unchecked exception (e.g. `RuntimeException`) to avoid forcing the user handling it explicitly every time

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            if (gitHubTokenOrganization == null) {
+                gitHubBuilder.withOAuthToken(gitHubToken);
+            } else {
+                gitHubBuilder.withOAuthToken(gitHubToken, gitHubTokenOrganization);
+            }
+        }
+        try {
+            gitHub = gitHubBuilder.build();
+        } catch (IOException ioException) {
+            logger.error("Failed to create GitHub client: {}", ioException.getMessage(), ioException);
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: no need to log if the exception is re-thrown further

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    public Document loadDocument(String owner, String repo, String branch, String path, DocumentParser parser) throws IOException {
+        GHContent content = null;
+        try {
+            content = gitHub
+                    .getRepository(owner + "/" + repo)
+                    .getFileContent(path, branch);
+        } catch (IOException ioException) {
+            logger.error("Failed to read GitHub repository: {}", ioException.getMessage(), ioException);
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: no need to log if the exception is re-thrown further

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 17]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    private static Document fromGitHub(DocumentParser parser, GHContent content) {
+        logger.info("Loading document from GitHub: {}", content.getHtmlUrl());
+        try {
+            if (content.isFile()) {
+                GitHubSource source = new GitHubSource(content);
+                return DocumentLoader.load(source, parser);
+            } else {
+                logger.error("Content must be a file, and not a directory: {}", content.getHtmlUrl());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: no need to log if the exception is re-thrown further

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+                throw new IllegalArgumentException("Content must be a file, and not a directory: " + content.getHtmlUrl());
+            }
+        } catch (IOException ioException) {
+            logger.error("Failed to load document from GitHub: {}", content.getHtmlUrl(), ioException);
+            return null;
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: If a single document failed to be loaded, we should propagate an exception to the caller. Returning null here seems wrong to me.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        public Builder gitHubTokenOrganization(String gitHubTokenOrganization) {
+            this.gitHubTokenOrganization = gitHubTokenOrganization;
+            return this;
+        }
+
+        public GitHubDocumentLoader build() throws IOException {
+            if (apiUrl == null) {
+                return new GitHubDocumentLoader(gitHubToken, gitHubTokenOrganization);
+            } else {
+                return new GitHubDocumentLoader(apiUrl, gitHubToken, gitHubTokenOrganization);
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: no need to check `apiUrl`, you can leave just this line (`else `branch), it is safe to pass `apiUrl=null`

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GitHubSource.java [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+    public Metadata metadata() {
+        Metadata metadata = new Metadata();
+        metadata.add("github_git_url", content.getGitUrl());
+        try {
+            metadata.add("github_download_url", content.getDownloadUrl());
+        } catch (IOException e) {
+            // Ignore if download_url is not available
+        }
+        metadata.add("github_html_url", content.getHtmlUrl());
+        metadata.add("github_name", content.getName());
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/source/github/GitHubSource.java]: Would be nice to make key names more specific, as I suggested earlier (e.g. "github_file_name", "github_file_size", etc).
"github_name", "github_size", etc sounds too ambiguous

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+    public List<Document> loadDocuments(String owner, String repo, String branch, String path, DocumentParser parser) throws IOException {
+        List<Document> documents = new ArrayList<>();
+        try {
+            gitHub
+                    .getRepository(owner + "/" + repo)
+                    .getDirectoryContent(path, branch)
+                    .forEach(ghDirectoryContent -> GitHubDocumentLoader.scanDirectory(ghDirectoryContent, documents, parser));
+        } catch (IOException ioException) {
+            logger.error("Failed to read GitHub repository: {}", ioException.getMessage(), ioException);
+            throw ioException;
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: I had an impression that the idea was to skip documents that fail to be loaded. If yes, then we should not throw here, just log

File: document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoaderIT.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+        assertThat(document.metadata("github_git_url")).startsWith("https://api.github.com/repos/langchain4j/langchain4j");
+    }
+
+    @Test
+    public void should_load_repository() throws IOException {
+        List<Document> documents = loader.loadDocuments(TEST_OWNER, TEST_REPO, "main", parser);
+
+        assertThat(documents.size()).isGreaterThan(1);
+    }
+}
```
Comment (langchain4j) [inline document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoaderIT.java:71]: It is hard to follow recursive logic intertwined with `for` loops and exception handling, so it would be awesome to have a few negative cases here (e.g. apparently wrong owner/repo/branch/path) for both single and multiple documents, to make sure we handle them properly

**Review by ghost — COMMENTED** [review]
@jdubois thanks a lot! Please check comments, few issues are still open

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java (hunk 14 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: Oh I thought you wanted to keep them, no problem let me refactor this.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java (hunk 15 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: I thought you wanted a log message, to help people understand what's going on? No problem I'll remove them.

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java (hunk 21 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: Yes, that's what is happening in the "scanDirectory" method, here the IOException is thrown by the `getRepository` method, and this one would be an exception (if you can get the repo, it's really bad!)

File: document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoaderIT.java (hunk 22 above)
Comment (jdubois) [reply document-loaders/langchain4j-document-loader-github/src/test/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoaderIT.java:71]: I added a couple of tests, but it's always the same simple exception, you get a GHFileNotFoundException when the file can't be found, so it's pretty simple.

**Comment by jdubois** [comment]
Thanks again @langchain4j , it should be OK now!

File: document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java (hunk 21 above)
Comment (langchain4j) [reply document-loaders/langchain4j-document-loader-github/src/main/java/dev/langchain4j/data/document/loader/github/GitHubDocumentLoader.java]: Got it, thanks for explanation!

**Review by ghost — APPROVED** [review]
@jdubois looks great, thank you so much!

**Comment by amuniz** [comment]
This is really useful, thanks! Could we get this into a release some time soon please @langchain4j?

**Comment by langchain4j** [comment]
Yes, release is planned on monday-tuesday. Following releases will become more frequent.

**Comment by jdubois** [comment]
I'm super interested by feedback and use cases @amuniz so don't hesitate to ping me!

**Comment by amuniz** [comment]
I'm writing an assistant that auto-generates a Jenkins pipeline for a given project in a GitHub repository. I wanted to use this loader in order to give the AI the required information so it knows what's the right build tool to use and what's the right call to the build tool.

**Comment by jdubois** [comment]
Very interesting, thank you! If you need some customization for this loader, we can work on this together.
